File: common/src/main/java/allyouneed/util/BigKeyCounter.java
package allyouneed.util;

import appeng.api.stacks.AEKey;
import appeng.api.stacks.KeyCounter;
import it.unimi.dsi.fastutil.objects.Object2ObjectMap;
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
import org.jetbrains.annotations.Nullable;

import java.math.BigInteger;
import java.util.Collections;
import java.util.Objects;
import java.util.Set;

/**
 * Network-wide amount tally using {@link BigInteger} to avoid long overflow when summing cells.
 */
public final class BigKeyCounter {
    private final Object2ObjectOpenHashMap<AEKey, BigInteger> amounts = new Object2ObjectOpenHashMap<>();

    public void clear() {
        amounts.clear();
    }

    public void add(AEKey key, long amount) {
        if (amount == 0) {
            return;
        }
        add(key, BigInteger.valueOf(amount));
    }

    public void add(AEKey key, BigInteger amount) {
        Objects.requireNonNull(key, "key");
        Objects.requireNonNull(amount, "amount");
        if (amount.signum() == 0) {
            return;
        }
        amounts.merge(key, amount, BigInteger::add);
    }

    public void addAll(KeyCounter other) {
        for (var entry : other) {
            add(entry.getKey(), entry.getLongValue());
        }
    }

    public void addAll(BigKeyCounter other) {
        for (var entry : other.amounts.object2ObjectEntrySet()) {
            add(entry.getKey(), entry.getValue());
        }
    }

    public void set(AEKey key, BigInteger amount) {
        Objects.requireNonNull(key, "key");
        if (amount == null || amount.signum() == 0) {
            amounts.remove(key);
        } else {
            amounts.put(key, amount);
        }
    }

    public BigInteger get(AEKey key) {
        Objects.requireNonNull(key, "key");
        return amounts.getOrDefault(key, BigInteger.ZERO);
    }

    public long getSaturatedLong(AEKey key) {
        return CommonKt.saturateToLong(get(key));
    }

    public boolean isEmpty() {
        return amounts.isEmpty();
    }

    public int size() {
        return amounts.size();
    }

    public Set<AEKey> keySet() {
        return Collections.unmodifiableSet(amounts.keySet());
    }

    public Object2ObjectMap.FastEntrySet<AEKey, BigInteger> entries() {
        return amounts.object2ObjectEntrySet();
    }

    public void removeZeros() {
        amounts.object2ObjectEntrySet().removeIf(e -> e.getValue().signum() == 0);
    }

    /**
     * Diff keys whose amounts differ from {@code other} (including keys only present on one side).
     */
    public void collectChangedKeys(BigKeyCounter other, java.util.function.Consumer<AEKey> out) {
        for (var entry : amounts.object2ObjectEntrySet()) {
            if (entry.getValue().compareTo(other.get(entry.getKey())) != 0) {
                out.accept(entry.getKey());
            }
        }
        for (var entry : other.amounts.object2ObjectEntrySet()) {
            if (!amounts.containsKey(entry.getKey())) {
                out.accept(entry.getKey());
            }
        }
    }

    public BigKeyCounter copy() {
        BigKeyCounter copy = new BigKeyCounter();
        copy.amounts.putAll(this.amounts);
        return copy;
    }

    /**
     * Add saturated long amounts into a KeyCounter for AE2 compatibility paths.
     */
    public void copySaturatedTo(KeyCounter out) {
        for (var entry : amounts.object2ObjectEntrySet()) {
            out.add(entry.getKey(), CommonKt.saturateToLong(entry.getValue()));
        }
    }

    @Nullable
    public static BigKeyCounter fromKeyCounter(KeyCounter counter) {
        if (counter == null) {
            return null;
        }
        BigKeyCounter big = new BigKeyCounter();
        big.addAll(counter);
        return big;
    }
}

File: common/src/main/kotlin/allyouneed/util/bigint/BigKeyCounter.kt
package allyouneed.util.bigint

import allyouneed.util.saturateToLong
import appeng.api.stacks.AEKey
import appeng.api.stacks.KeyCounter
import it.unimi.dsi.fastutil.objects.Object2ObjectMap
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap
import java.math.BigInteger
import java.util.*
import java.util.function.Consumer

/** Network-wide amount tally using [BigInteger] to avoid long overflow when summing cells. */
class BigKeyCounter {
    private val amounts = Object2ObjectOpenHashMap<AEKey, BigInteger>()

    fun clear() {
        amounts.clear()
    }

    fun add(key: AEKey, amount: Long) {
        if (amount == 0L) return
        add(key, BigInteger.valueOf(amount))
    }

    fun add(key: AEKey, amount: BigInteger) {
        if (amount.signum() == 0) return
        amounts.merge(key, amount, BigInteger::add)
    }

    fun addAll(other: KeyCounter) {
        for (entry in other) {
            add(entry.key, entry.longValue)
        }
    }

    fun addAll(other: BigKeyCounter) {
        for (entry in other.amounts.object2ObjectEntrySet()) {
            add(entry.key, entry.value)
        }
    }

    fun set(key: AEKey, amount: BigInteger?) {
        if (amount == null || amount.signum() == 0) {
            amounts.remove(key)
        } else {
            amounts[key] = amount
        }
    }

    fun get(key: AEKey): BigInteger = amounts.getOrDefault(key, BigInteger.ZERO)

    fun getSaturatedLong(key: AEKey): Long = get(key).saturateToLong()

    fun isEmpty(): Boolean = amounts.isEmpty()

    fun size(): Int = amounts.size

    fun keySet(): Set<AEKey> = Collections.unmodifiableSet(amounts.keys)

    fun entries(): Object2ObjectMap.FastEntrySet<AEKey, BigInteger> = amounts.object2ObjectEntrySet()

    fun removeZeros() {
        amounts.object2ObjectEntrySet().removeIf { it.value.signum() == 0 }
    }

    /** Diff keys whose amounts differ from [other] (including keys only present on one side). */
    fun collectChangedKeys(other: BigKeyCounter, out: Consumer<AEKey>) {
        for (entry in amounts.object2ObjectEntrySet()) {
            if (entry.value.compareTo(other.get(entry.key)) != 0) {
                out.accept(entry.key)
            }
        }
        for (entry in other.amounts.object2ObjectEntrySet()) {
            if (!amounts.containsKey(entry.key)) {
                out.accept(entry.key)
            }
        }
    }

    fun copy(): BigKeyCounter {
        val copy = BigKeyCounter()
        copy.amounts.putAll(amounts)
        return copy
    }

    /** Add saturated long amounts into a KeyCounter for AE2 compatibility paths. */
    fun copySaturatedTo(out: KeyCounter) {
        for (entry in amounts.object2ObjectEntrySet()) {
            out.add(entry.key, entry.value.saturateToLong())
        }
    }

    companion object {
        @JvmStatic
        fun fromKeyCounter(counter: KeyCounter?): BigKeyCounter? {
            if (counter == null) return null
            return BigKeyCounter().also { it.addAll(counter) }
        }
    }
}
